examples/hal/compute/main.rs

```diff
@@ -0,0 +1,162 @@
+extern crate gfx_backend_vulkan as back;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "gl")]
+extern crate gfx_backend_gl as back;
+#[cfg(feature = "gl")]
+extern crate glutin;
+#[cfg(feature = "gl")]
+use glutin::HeadlessRendererBuilder;
+
+use std::str::FromStr;
+
+use hal::{Backend, Adapter, Gpu, Device, DescriptorPool, Instance};
+use hal::{queue, pso, memory, buffer, pool, command, device};
+
+#[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal", feature = "gl"))]
+fn main() {
+    env_logger::init();
+
+    #[cfg(any(feature = "vulkan", feature = "dx12"))]
+    let instance = back::Instance::create("gfx-rs compute", 1);
+
+    #[cfg(feature = "gl")]
+    let instance = back::Headless(HeadlessRendererBuilder::new(256, 256).build().unwrap()); // dimensions?
+
+    #[cfg(feature = "metal")]
+    let instance = back::Instance::create();
+
+    let mut gpu = instance.enumerate_adapters().into_iter()
+        .find(|a| a.get_queue_families().into_iter().any(|&(_, t)| t.supports_compute()))
+        .expect("Failed to find a GPU with compute support!")
+        .open_with(|_, qtype| {
+            if qtype.supports_compute() {
+                (1, queue::QueueType::Compute)
+            }
+            else {
+                (0, queue::QueueType::Compute)
+            }
+        });
+    
+    #[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal"))]
+    let shader = gpu.device.create_shader_module(include_bytes!("shader/collatz.spv")).unwrap();
+        
+    #[cfg(feature = "gl")]
+    let shader = gpu.device.create_shader_module_from_source(include_bytes!("shader/collatz.comp"), pso::Stage::Compute).unwrap();
+
+    let (pipeline_layout, pipeline, set_layout, mut desc_pool) = {
+        let set_layout = gpu.device.create_descriptor_set_layout(&[
+                pso::DescriptorSetLayoutBinding {
+                    binding: 0,
+                    ty: pso::DescriptorType::StorageBuffer,
+                    count: 1,
+                    stage_flags: pso::STAGE_COMPUTE,
+                }
+            ],
+        );
+
+        let pipeline_layout = gpu.device.create_pipeline_layout(&[&set_layout]);
+        let entry_point = pso::EntryPoint { entry: "main", module: &shader };
+        let pipeline = gpu.device.create_compute_pipelines(&[(entry_point, &pipeline_layout)]).remove(0).expect("Error creating compute pipeline!");
+
+        let desc_pool = gpu.device.create_descriptor_pool(
+            1,
+            &[
+                pso::DescriptorRangeDesc {
+                    ty: pso::DescriptorType::StorageBuffer,
+                    count: 1,
+                },
+            ],
+        );
+        (pipeline_layout, pipeline, set_layout, desc_pool)
+    };
+
+    // For now this just panics if you didn't pass numbers. Could add proper error handling.
+    let numbers: Vec<u32> = std::env::args().skip(1).map(|s| u32::from_str(&s).unwrap()).collect();
```

Comment:
Move this to the beginning imo to fail early. And give at least an error message on panic to let the user know how to run the program.